service/src/main/java/uk/nhs/adaptors/gp2gp/ehr/request/EhrExtractAckHandler.java

```diff
@@ -0,0 +1,84 @@
+    private static final String MESSAGE_ID_ROOT_XPATH = "//MCCI_IN010000UK13/id/@root";
+    private static final String ERROR_CODE_XPATH = "//justifyingDetectedIssueEvent/code";
+    private static final String ACK_OK_CODE = "AA";
+    private static final String ACK_BUSINESS_ERROR_CODE = "AE";
+    private static final String ACK_REJECTED_CODE = "AR";
+    private static final String CODE_ATTRIBUTE = "code";
+    private static final String DISPLAY_ATTRIBUTE = "displayName";
+
+
+    private final XPathService xPathService;
+    private final EhrExtractStatusService ehrExtractStatusService;
+
+    @SneakyThrows
+    public void handle(String conversationId, Document document) {
+        String ackTypeCode = xPathService.getNodeValue(document, ACK_TYPE_CODE_XPATH);
+        Instant now = Instant.now();
+        String messageRef = xPathService.getNodeValue(document, MESSAGE_REF_XPATH);
+        String rootId = xPathService.getNodeValue(document, MESSAGE_ID_ROOT_XPATH);
+        EhrReceivedAcknowledgementBuilder ackBuilder = EhrReceivedAcknowledgement.builder()
+                .messageRef(messageRef)
+                .rootId(rootId)
+                .received(now);
+
+        if (ACK_OK_CODE.equals(ackTypeCode)) {
+            LOGGER.info("Application Acknowledgement Accept ({}) received, messageRef: {}", messageRef, ackTypeCode);
+            ackBuilder.conversationClosed(now);
+        } else if (ACK_BUSINESS_ERROR_CODE.equals(ackTypeCode)) {
+            LOGGER.info("Application Acknowledgement Error ({}) received, messageRef: {}", messageRef, ackTypeCode);
+            ackBuilder.errors(extractErrorCodes(document, ERROR_CODE_XPATH));
+        } else if (ACK_REJECTED_CODE.equals(ackTypeCode)) {
+            LOGGER.info("Application Acknowledgement Reject ({}) received, messageRef: {}", messageRef, ackTypeCode);
+            ackBuilder.errors(extractErrorCodes(document, ACK_DETAILS_XPATH));
+        } else {
+            throw new InvalidInboundMessageException("Unsupported MCCI_IN010000UK13/acknowledgement/@typeCode: " + ackTypeCode);
+        }
+
+        ehrExtractStatusService.updateEhrExtractStatusAck(conversationId, ackBuilder.build());
+    }
+
+    private List<ErrorDetails> extractErrorCodes(Document document, String xPath) {
+        NodeList errors = xPathService.getNodes(document, xPath);
+        List<ErrorDetails> errorDetailsList = new ArrayList<>();
+        for (int i = 0; i < errors.getLength(); i++) {
+            Node error = errors.item(i);
+            ErrorDetails errorDetails = ErrorDetails.builder()
+                    .code(error.getAttributes().getNamedItem(CODE_ATTRIBUTE).getNodeValue())
+                    .display(error.getAttributes().getNamedItem(DISPLAY_ATTRIBUTE).getNodeValue())
+                    .build();
+
+            errorDetailsList.add(errorDetails);
+        }
+
+        return errorDetailsList;
```

Comment:
This is nostalgic! 😁

How about:
```suggestion
        return IntStream.range(0, errors.getLength())
            .mapToObj(errors::item)
            .map(error -> ErrorDetails.builder()
                .code(error.getAttributes().getNamedItem(CODE_ATTRIBUTE).getNodeValue())
                .display(error.getAttributes().getNamedItem(DISPLAY_ATTRIBUTE).getNodeValue())
                .build())
            .collect(Collectors.toList());
```

Reply:
I think that good old for loop is sometimes acceptable. Especially when you're dealing with such an old API like XPath. I don't mind the new approach, happy to accept it ;)